torch/nn/functional.py

```diff
@@ -4916,11 +4916,19 @@ def _in_projection(
 
     # Efficient implementation equivalent to the following:
     scale_factor = 1 / math.sqrt(Q.size(-1)) if scale is None else scale
-    attn_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0) if is_causal else attn_mask
-    attn_mask = attn_mask.masked_fill(not attn_mask, -float('inf')) if attn_mask.dtype==torch.bool else attn_mask
-    attn_weight = torch.softmax((Q @ K.transpose(-2, -1) * scale_factor) + attn_mask, dim=-1)
-    attn_weight = torch.dropout(attn_weight, dropout_p)
-    return attn_weight @ V
+    if is_causal:
+        assert attn_mask is None
+        attn_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0)
+    attn_bias = torch.zeros(L, S, dtype=Q.dtype)
```

Comment:
to make this runnable as @vadimkantorov mentioned, perhaps define `L = ...` and `S = ...` above?

Reply:
I think it's best to have this reference code as function. This would make clearer to add a return statement at the end and will make clearer the arguments